validator/client/attest_protect.go

```diff
@@ -26,46 +28,41 @@ func (v *validator) slashableAttestationCheck(
 	ctx, span := trace.StartSpan(ctx, "validator.postAttSignUpdate")
 	defer span.End()
 
-	fmtKey := fmt.Sprintf("%#x", pubKey[:])
+	// Based on EIP3076, validator should refuse to sign any attestation with source epoch less
+	// than the minimum source epoch present in that signer’s attestations.
+	lowestSourceEpoch, exists, err := v.db.LowestSignedSourceEpoch(ctx, pubKey)
+	if err != nil {
+		return err
+	}
+	if exists && indexedAtt.Data.Source.Epoch < lowestSourceEpoch {
+		return fmt.Errorf(
+			"could not sign attestation lower than lowest source epoch in db, %d < %d",
+			indexedAtt.Data.Source.Epoch,
+			lowestSourceEpoch,
+		)
+	}
+	existingSigningRoot, err := v.db.SigningRootAtTargetEpoch(ctx, pubKey, indexedAtt.Data.Target.Epoch)
+	if err != nil {
+		return err
+	}
+	signingRootsDiffer := slashutil.SigningRootsDiffer(existingSigningRoot, signingRoot)
+
+	// Based on EIP3076, validator should refuse to sign any attestation with target epoch less
+	// than or equal to the minimum target epoch present in that signer’s attestations.
+	lowestTargetEpoch, exists, err := v.db.LowestSignedTargetEpoch(ctx, pubKey)
+	if err != nil {
+		return err
+	}
+	if signingRootsDiffer && exists && indexedAtt.Data.Target.Epoch <= lowestTargetEpoch {
+		return fmt.Errorf(
+			"could not sign attestation lower than or equal to lowest target epoch in db, %d <= %d",
+			indexedAtt.Data.Target.Epoch,
+			lowestTargetEpoch,
+		)
+	}
+	fmtKey := hex.EncodeToString(pubKey[:])
```

Comment:
Unfortunately, the previous sprintf was using 0x prefix so you're going to have to prepend this key or do some database migration.

https://play.golang.org/p/xtnT4tvSma5

Reply:
oh wait, it's just for logging? maybe want to prepend anyway

Reply:
yep for logging, will prepend